src/process-thread-coroutine.md

```diff
@@ -0,0 +1,164 @@
+section, one x86 CPU can at most run 128 tasks at the same time, [AMD Ryzen™
+Threadripper™ PRO 5995WX Processor][2]. But it's too easy to create thousands of
+processes or threads on Linux, we have to decide how to place them on the CPU
+core and when to stop a task, where [OS task scheduler][3] helps.
+
+Schedulers can interrupt an executing task regardless of its state, and schedule
+a new one. It's called preemptive schedule and is used by most OSes like Linux. The
+advantage is that it can share the CPU time slice between tasks fairly no matter
+what they're running, but the tasks have no idea about the scheduler. To
+interrupt a running task, hardware interruption like time interruption is
+necessary.
+
+The other schedulers are called non-preemptive schedulers, which have to
+cooperate with the task while scheduling. Here tasks are not interrupted,
+instead, they decide when to release the computing resource. The tasks usually 
+schedule themselves out when doing I/O operations, which usually take a while
+to complete. Fairness is hard to be guaranteed as the task itself may run
+forever without stopping, in which case other tasks have no opportunity to be
+scheduled on that core.
+
+No matter what kind of scheduler is taken, tasks scheduling always needs to do the
+following steps:
+
+* Store current process/thread execution flow information.
+* Change page table mapping (memory space) and flush TLB if necessary.
+* Restore the new process/thread execution flow from the previously stored
+  state.
+
+After adopting a scheduler operating system can run tens of thousands of
+processes/threads on the limited hardware resource.
+
+# Coroutine
+
+We have basic knowledge of OS scheduling, and it seems to work fine in most
+cases. Next, let's see how it performs in extreme scenarios. Free software
+developer, [Jim Blandy][4], did an [interesting test][5] to show how much time
+it takes to do a context switch on Linux. In the test, the app creates 500
+thread and connect them with pipes like a chain, and then pass a one-byte
+message from one side to the other side. The whole test runs 10000 iterations to
+get a stable result. The result shows that a thread context switch takes around
+1.7us, compared to 0.2us of a Rust async task switch.
+
+It's the first time to mention "Rust async task", formally it's called
+[coroutine][6]. The coroutines are lightweight tasks for non-preemptive
+multitasking, whose execution can be suspended and resumed. Usually, the task
+itself decides when to suspend and wait for a notification to resume. To suspend
+and resume tasks' execution flow, the execution states should be saved, just
+like what OS does. Saving the CPU register values is easy for the OS, but not
+for the applications. Rust saves it to a state machine, and the machine can only
+be suspended and resumed from the valid states in that machine. We name the
+state machine "Future".
```

Comment:
The underlying state machine should be ` generator`, right?

Reply:
`Future` is an abstraction built on `generator`. It is better to have a clear hierarchy, otherwise it is easy to confuse the reader.

Reply:
Yes, future is an interface, and the `generator` implement the interface.